scripts/plugins/extauth-hook-AD.py

```diff
@@ -9,138 +9,268 @@
         else:
-            os.write(self.temp_fd, "account sufficient pam_succeed_if.so user = %s\n" % name)
+            content = self.no_ad_pam
+        lines = content.split("\n")
+        self._lines = ["{}\n".format(l) for l in lines[:-1]] + lines[-1:] # Do not add \n to last element
 
+class DynamicPam(ADConfig):
+    def __init__(self, session, arg, ad_enabled=True):
+        super(DynamicPam, self).__init__("/etc/pam.d/hcp_users", session, arg, ad_enabled)
+        self.lines = []
 
-def rewrite_etc_pamd_ssh(session, args):
-    # Rewrite the PAM SSH config using the latest info from Active Directory
-    # and the list of subjects from xapi
-    try:
-        config = PamSshConfig()
-        subjects = session.xenapi.subject.get_all()
-        admin_role = session.xenapi.role.get_by_name_label('pool-admin')[0]
+    def _apply_to_cache(self):
+        # AD is not enabled, just return as the configure file will be removed during install
+        if not self._ad_enabled:
+            return
+        # Rewrite the PAM SSH config using the latest info from Active Directory
+        # and the list of subjects from xapi
+        subjects = self._session.xenapi.subject.get_all()
+        admin_role = self._session.xenapi.role.get_by_name_label('pool-admin')[0]
         # Add each subject which contains the admin role
         for opaque_ref in subjects:
-            subject_rec = session.xenapi.subject.get_record(opaque_ref)
+            subject_rec = self._session.xenapi.subject.get_record(opaque_ref)
             sid = subject_rec['subject_identifier']
+            name = subject_rec["other_config"]["subject-name"]
+            is_group = True if subject_rec["other_config"]["subject-is-group"] is "true" else False
+            logger.debug("Permit %s with sid %s is_group as %s", name, sid, is_group)
             if admin_role in subject_rec['roles']:
-                config.add_subject(sid)
-        config.install()
+                self._add_subject(name, is_group)
+
+    def _add_subject(self, name, is_group):
+        condition = " ingroup {}".format(name) if is_group else " = {}".format(name)
+        self._lines.append("account sufficient pam_succeed_if.so user {} \n".format(condition))
+
+    def _install(self):
+        if self._ad_enabled:
+            super(DynamicPam, self)._install()
+        else:
+            if os.path.exists(self._file_path):
+                os.remove(self._file_path)
+
+
+class KeyValueConfig(ADConfig):
+    def __init__(self, path, session, args, ad_enabled=True, sep=": "):
+        super(KeyValueConfig, self).__init__(path, session, args, ad_enabled)
+        self._sep = sep
+        self._values = {}
+
+    def _update_value(self, key, value):
+        self._values[key] = value
+
+    def _apply_value(self, key, value):
+        try:
+            target_line = "{}{}{}\n".format(key, self._sep, value)
+            for idx, line in enumerate(self._lines):
```

Comment:
instead of using a list of encoded strings to store the key value pairs, we should really use a dictionary

Reply:
The key value pairs do stored in dict in `self._values`,   

the `self._lines` are raw line read from original existing configuration file.

Reply:
> The key value pairs do stored in dict in self._values,

Indeed, you have `def _update_value`, so why do we need `def _apply_value` as well? It's like you are storing key values pairs in both `self._values` and in `self._lines` which I don't understand

Reply:
`self._lines` is general for all kinds of configurations,  (not only for key value pair configuration),    It is used as cache (which will be write to file as configuration).  It will be initialized with file existing content on demand. 

`self._values` is only specific for key value pair configurations,  as you said, it is a dict to store key, values. 
Since KeyValueCOnfig needs to reuse the parent function like `_install`, so it needs to update its value to `self._lines` on its `apply()`

Reply:
in that case can we store the 'key value pair configuration' separately from the 'general configuration'? I think that would simplify this code here

> Since KeyValueCOnfig needs to reuse the parent function like _install, so it needs to update its value to self._lines on its apply()

I think it would be simpler to have the install function generate the string from an intermediate representation, rather than having to maintain/update the line strings

Reply:
> in that case can we store the 'key value pair configuration' separately from the 'general configuration'? I think that would simplify this code here
> 
> > Since KeyValueCOnfig needs to reuse the parent function like _install, so it needs to update its value to self._lines on its apply()
> 
They are stored separately from the `general configuration`, let's have a talk offline. 
> I think it would be simpler to have the install function generate the string from an intermediate representation, rather than having to maintain/update the line strings